models/migrations/v229.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"fmt"
+
+	"code.gitea.io/gitea/models/webhook"
+	"code.gitea.io/gitea/modules/json"
+	"code.gitea.io/gitea/modules/secret"
+	"code.gitea.io/gitea/modules/setting"
+	api "code.gitea.io/gitea/modules/structs"
+	"code.gitea.io/gitea/modules/timeutil"
+
+	"xorm.io/builder"
+	"xorm.io/xorm"
+)
+
+func batchProcess[T any](x *xorm.Engine, query func() *xorm.Session, buf []T, process func(*xorm.Session, T) error) error {
+	size := cap(buf)
+	start := 0
+	for {
+		err := query().Limit(size, start).Find(&buf)
+		if err != nil {
+			return err
+		}
+		if len(buf) == 0 {
+			return nil
+		}
+
+		err = func() error {
+			sess := x.NewSession()
+			defer sess.Close()
+			if err := sess.Begin(); err != nil {
+				return fmt.Errorf("unable to allow start session. Error: %w", err)
+			}
+			for _, record := range buf {
+				if err := process(sess, record); err != nil {
+					return err
+				}
+			}
+			return sess.Commit()
+		}()
+		if err != nil {
+			return err
+		}
+
+		if len(buf) < size {
+			return nil
+		}
+		start += size
+		buf = buf[:0]
+	}
```

Comment:
```suggestion
	for len(buf) < size {
		err := query().Limit(size, start).Find(&buf)
		if err != nil {
			return err
		}
		if len(buf) == 0 {
			return nil
		}
		
		if err = db.WithTx(func(ctx context.Context) error {
			for _, record := range buf {
				if err := process(sess, record); err != nil {
					return err
				}
			}
			return nil
		}); err != nil {
			return err
		}

		start += size
		buf = buf[:0]
	}
	return nil
```
(I hope I produced valid Go code?)

Reply:
Could you provide a bit more background on your change suggestion ?

At line `if err := process(sess, record); err != nil {`, `sess` is now undefined. I guess it should be replaced by `ctx`.

However I have no clue how `sess.ID(...).Cols(...).Update(...)` should be updated to work with `ctx`. `models/db/context.go` does provide `Insert` and `Delete`, but no `Update` method :interrobang: